fix(sem): analysis considers nested noreturn

saem opened the pull request:
## Summary
No-return analysis ensures that nested final expressions (
`block/case/if/try` ) are accounted for no-return itself.

## Details
Prior to this change if the last expression within a greater expression
was a no-return expression it would not be deemed as such if it was a 
`block/case/if/try` , instead errors would be raised related to discard
checks or that path not resulting in the correct type.

For example, the following would result in an error where  `0`  must be
used or discarded:
```nim
proc foo() =
  let x =
    if false: 10
    else:
      block:
        return
```

Fixes https://github.com/nim-works/nimskull/issues/1440

saem committed:
noreturn analysis accounts for `if/case/try/block`

saem committed:
fix bad formatting, it was really annoying

saem commented on compiler/ast/ast_query.nim:
```diff
@@ -690,6 +690,16 @@ iterator genericParamsInMacroCall*(macroSym: PSym, call: PNode): (PSym, PNode) =
 
+proc endsInNoReturn*(n: PNode): bool =
+  ## checks if expression `n` ends in an unstructured exit (raise, return,
+  ## etc), calls of noreturn proc, or
```
```suggestion
  ## etc) or calls of noreturn proc. This is meant to be called on an semmed `n`.
```

saem committed:
complete doc comment

zerbina commented on compiler/ast/ast_query.nim:
```diff
@@ -690,6 +690,16 @@ iterator genericParamsInMacroCall*(macroSym: PSym, call: PNode): (PSym, PNode) =
       yield (genericParam, call[posInCall])
 
+proc endsInNoReturn*(n: PNode): bool =
+  ## checks if expression `n` ends in an unstructured exit (raise, return,
+  ## etc) or calls of noreturn proc. This is meant to be called on an semmed `n`.
```
```suggestion
  ## Checks if expression `n` ends in an unstructured exit (raise, return,
  ## etc.) or a call of a noreturn proc. This is meant to be called on a
  ## semmed `n`.
```
Fixes a few small issues (`an` instead of `a`, plural for call, etc.).

zerbina commented on tests/lang_exprs/tnoreturn_nested.nim:
```diff
@@ -0,0 +1,53 @@
+discard """
+  description: "
+    Ensure nested noreturn statements are considererd in noreturn analysis
+  "
```
```suggestion
  description: """
    Ensure nested noreturn statements are considererd in noreturn analysis
  """
```

zerbina commented on compiler/ast/ast_query.nim:
```diff
@@ -690,6 +690,16 @@ iterator genericParamsInMacroCall*(macroSym: PSym, call: PNode): (PSym, PNode) =
+  while it.kind in {nkStmtList, nkStmtListExpr} and it.len > 0:
+    it = it.lastSon
+  result = it.kind in nkLastBlockStmts or
+    it.kind in {nkIfStmt, nkIfExpr, nkTryStmt, nkCaseStmt, nkTryStmt, nkBlockExpr} and it.typ.isNil or
```
```suggestion
    (it.kind in {nkIfStmt, nkTryStmt, nkCaseStmt, nkBlockStmt} and
     it.typ.isEmptyType()) or
```
`nkIfExpr` and `nkBlockExpr` don't need to be handled here, since they're guaranteed to have a non-`void` type.

Some noreturn compound statements can also have a non-nil `typ` (when assigned the `enforceVoidContext` type), so `isEmptyType` (which returns true for both `typ == nil` and `typ.kind == tyVoid`) has to be used.

saem commented on tests/lang_exprs/tnoreturn_nested.nim:
```diff
@@ -0,0 +1,53 @@
+      0
+    else:
+      let s = false
+      case
```
```suggestion
      case s
```

saem committed:
Apply suggestions from code review

- quoting test description used double quotes, need triple-double-quotes
  - surfaced a testament bug that silently ignores error on test joining
- fix nested `case` test
- clean-up `endsInNoReturn` docs
- fix nested checking, both checking `.typ` field with `isEmptyType` and focus on stmt variants of nodes

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

zerbina commented on tests/lang_exprs/tnoreturn_nested.nim:
```diff
@@ -0,0 +1,53 @@
+discard """
+  description: """
+    Ensure nested noreturn statements are considererd in noreturn analysis
+  """
```
```suggestion
  description: '''
    Ensure nested noreturn statements are considererd in noreturn analysis
  '''
```
My bad, it needs to be triple single-quotes.

saem committed:
Update tests/lang_exprs/tnoreturn_nested.nim

Co-authored-by: zerbina <100542850+zerbina@users.noreply.github.com>

saem commented:
/merge
